File: tests/AqiClock.Domain.Tests/Scheduling/TestData.cs
using System.Globalization;
using AqiClock.Domain.Entities;
using AqiClock.Domain.Scheduling;

namespace AqiClock.Domain.Tests.Scheduling;

/// <summary>Builders for concise test setup. Monday 2026-07-13 is the reference week.</summary>
internal static class TestData
{
    public static readonly DateOnly Monday = new(2026, 7, 13);
    public static readonly DateOnly Tuesday = new(2026, 7, 14);

    public static Period P(string name, string start, string end, int sort = 0, bool isLesson = true, Guid? id = null) =>
        new(id ?? Guid.NewGuid(),
            name,
            TimeOnly.Parse(start, CultureInfo.InvariantCulture),
            TimeOnly.Parse(end, CultureInfo.InvariantCulture),
            sort,
            isLesson);

    public static Timetable Tt(string name, params Period[] periods) =>
        new(Guid.NewGuid(), name, IsArchived: false, periods);

    /// <summary>A snapshot where every weekday (Mon–Fri) uses <paramref name="timetable"/>.</summary>
    public static ScheduleSnapshot WeekOf(Timetable timetable, params DateOverride[] overrides) =>
        new(
            [timetable],
            new WeekSchedule(new Dictionary<DayOfWeek, Guid?>
            {
                [DayOfWeek.Monday] = timetable.Id,
                [DayOfWeek.Tuesday] = timetable.Id,
                [DayOfWeek.Wednesday] = timetable.Id,
                [DayOfWeek.Thursday] = timetable.Id,
                [DayOfWeek.Friday] = timetable.Id,
            }),
            overrides);

    /// <summary>A standard three-period day: 09:00–10:00, 10:00–10:20 break, 10:20–11:20.</summary>
    public static Timetable NormalDay() =>
        Tt("Normal Day",
            P("Period 1", "09:00", "10:00", sort: 1),
            P("Break", "10:00", "10:20", sort: 2, isLesson: false),
            P("Period 2", "10:20", "11:20", sort: 3));

    public static DateTime At(DateOnly date, string time) =>
        date.ToDateTime(TimeOnly.Parse(time, CultureInfo.InvariantCulture));
}

File: mobile/test/testData.ts
import {
  DateOverride,
  Period,
  ScheduleSnapshot,
  Timetable,
  WeekSchedule,
} from '@/src/domain/scheduleTypes';

export const monday = localDate(2026, 7, 13);
export const tuesday = localDate(2026, 7, 14);

let nextId = 1;
export function id(): string {
  return `00000000-0000-0000-0000-${(nextId++).toString().padStart(12, '0')}`;
}

export function period(
  name: string,
  startTime: string,
  endTime: string,
  sortOrder = 0,
  isLesson = true,
  periodId = id(),
): Period {
  return { id: periodId, name, startTime, endTime, sortOrder, isLesson };
}

export function timetable(name: string, ...periods: Period[]): Timetable {
  return { id: id(), name, isArchived: false, periods };
}

export function normalDay(): Timetable {
  return timetable(
    'Normal Day',
    period('Period 1', '09:00', '10:00', 1),
    period('Break', '10:00', '10:20', 2, false),
    period('Period 2', '10:20', '11:20', 3),
  );
}

export function weekOf(item: Timetable, ...dateOverrides: DateOverride[]): ScheduleSnapshot {
  return {
    timetables: [item],
    weekSchedule: week([0, item.id], [1, item.id], [2, item.id], [3, item.id], [4, item.id]),
    dateOverrides,
  };
}

export function week(...entries: readonly [number, string | null][]): WeekSchedule {
  return entries.map(([weekday, timetableId], index) => ({ id: `week-${weekday}-${index}`, weekday, audienceClassId: null, timetableId }));
}

export function override(date: Date, timetableId: string | null): DateOverride {
  return { id: id(), date: dateKey(date), timetableId };
}

export function at(date: Date, time: string): Date {
  const [hours, minutes, seconds = 0] = time.split(':').map(Number);
  return new Date(date.getFullYear(), date.getMonth(), date.getDate(), hours, minutes, seconds);
}

export function localDate(year: number, month: number, day: number): Date {
  return new Date(year, month - 1, day);
}

export function dateKey(date: Date): string {
  return `${date.getFullYear().toString().padStart(4, '0')}-${(date.getMonth() + 1)
    .toString()
    .padStart(2, '0')}-${date.getDate().toString().padStart(2, '0')}`;
}
